core-bundle/src/Twig/Extension/ContaoExtension.php

```diff
@@ -0,0 +1,126 @@
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Extension;
+
+use Contao\CoreBundle\Twig\Inheritance\DynamicExtendsTokenParser;
+use Contao\CoreBundle\Twig\Inheritance\TemplateHierarchyInterface;
+use Contao\CoreBundle\Twig\Interop\ContaoEscaper;
+use Contao\CoreBundle\Twig\Interop\ContaoEscaperNodeVisitor;
+use Contao\CoreBundle\Twig\Interop\PhpTemplateProxyNodeVisitor;
+use Contao\Template;
+use Twig\Environment;
+use Twig\Extension\AbstractExtension;
+use Twig\Extension\EscaperExtension;
+use Webmozart\PathUtil\Path;
+
+class ContaoExtension extends AbstractExtension
+{
+    /**
+     * @var TemplateHierarchyInterface
+     */
+    private $hierarchy;
+
+    /**
+     * @var array
+     */
+    private $contaoEscaperFilterRules = [];
+
+    public function __construct(Environment $environment, TemplateHierarchyInterface $hierarchy)
+    {
+        /** @var EscaperExtension $escaperExtension */
+        $escaperExtension = $environment->getExtension(EscaperExtension::class);
+
+        $escaperExtension->setEscaper(
+            'contao_html',
+            [(new ContaoEscaper()), '__invoke']
+        );
+
+        $this->hierarchy = $hierarchy;
+
+        // Use our escaper on all templates in the `@Contao` and `@Contao_*`
+        // namespaces
+        $this->addContaoEscaperRule('%^@Contao(_[a-zA-Z0-9_-]*)?/%');
+    }
+
+    /**
+     * Add a contao escaper rule.
+     *
+     * If a template name matches any of the defined rules it will be processed
+     * with the `contao_html` escaper strategy. Make sure your rule will only
+     * match templates with input encoded contexts!
+     */
+    public function addContaoEscaperRule(string $regularExpression): void
+    {
+        if (\in_array($regularExpression, $this->contaoEscaperFilterRules, true)) {
+            return;
+        }
+
+        $this->contaoEscaperFilterRules[] = $regularExpression;
+    }
+
+    public function getNodeVisitors(): array
+    {
+        return [
+            // Enables the 'contao_twig' escaper for Contao templates with
+            // input encoding
+            new ContaoEscaperNodeVisitor(
+                function () {
+                    return $this->contaoEscaperFilterRules;
+                }
+            ),
+            // Allows rendering PHP templates with the legacy framework by
+            // installing proxy nodes
+            new PhpTemplateProxyNodeVisitor(self::class),
+        ];
+    }
+
+    public function getTokenParsers(): array
+    {
+        return [
+            // Registers a parser for the 'extends' tag which will overwrite
+            // the one of Twig's CoreExtension
+            new DynamicExtendsTokenParser($this->hierarchy),
+        ];
+    }
+
+    /**
+     * @see \Contao\CoreBundle\Twig\Interop\PhpTemplateProxyNode
+     * @see \Contao\CoreBundle\Twig\Interop\PhpTemplateProxyNodeVisitor
+     *
+     * @internal
+     */
```

Comment:
This function is called from within Twig templates if a Contao template should get rendered (there are similar constructs in the Symfony's Twig profiler extension). This basically calls the old template engine with prepopulated blocks and block names and returns the rendered ("parsed") result.